### PR TITLE
Add leash component to pAIs that keeps them within range instead of directly teleporting them back, increases default range to max range

### PR DESCRIPTION
**Pull request by Mothblocks**
## About The Pull Request

Tries to keep pAIs in range of their owner by moving them closer when the owner moves, rather than jarringly teleporting every time the owner gets out of range. Does this by calculating the closest path a nearby tile and forcefully moving you there. Still a bit janky at times but is better than teleporting directly onto the owner 100% of the time I feel. Also prevents you from moving out of range, rather than forcefully teleporting you back.

Increases the default pAI range to the maximum (9 tiles)

## Why It's Good For The Game

New leashing makes being a leashed pAI significantly less jarring and obvious. Ideally we would also have a visible max range too.

Default pAI range was pretty small in my testing and I think it's not unreasonable to think a lot of people won't bother changing it. That they are leashed at all is the important part.

## Changelog

:cl:
qol: pAIs now try to stay within range of their owner, and teleport back only when necessary
qol: Default max pAI range has been changed to the maximum range you can choose (9 tiles)
/:cl:

### CODE REVIEW

File: code/datums/components/leash.dm [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+
+/datum/component/leash/proc/force_teleport_back(reason)
+	PRIVATE_PROC(TRUE)
+
+	var/atom/movable/movable_parent = parent
+
+	SSblackbox.record_feedback("tally", "leash_force_teleport_back", 1, reason)
+
+	if (force_teleport_out_effect)
+		new /obj/effect/temp_visual/guardian/phase/out(movable_parent.loc)
```
Comment (Jacquerel) [inline code/datums/components/leash.dm:157]:
```suggestion
		new force_teleport_out_effect(movable_parent.loc)
```
`force_teleport_out_effect` is a path not a boolean, right?

File: code/datums/components/leash.dm [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+
+	movable_parent.forceMove(get_turf(owner))
+
+	if (force_teleport_in_effect)
+		new /obj/effect/temp_visual/guardian/phase(movable_parent.loc)
```
Comment (Jacquerel) [inline code/datums/components/leash.dm:162]:
```suggestion
		new force_teleport_in_effect(movable_parent.loc)
```
Same here

File: code/datums/components/leash.dm [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+
+	if (ismob(movable_parent))
+		movable_parent.balloon_alert(movable_parent, "moved out of range!")
+
+	SEND_SIGNAL(parent, COMSIG_LEASH_FORCE_TELEPORT)
+
+/// A debug spawner that will create a corgi leashed to a bike horn, plus a beam
+/obj/effect/spawner/debug_leash
```
Comment (Jacquerel) [inline code/datums/components/leash.dm:170]: We want to keep this?

**Review by Jacquerel — CHANGES_REQUESTED** [review]
Small things.

**Comment by TheVekter** [comment]
Yeah, thank you for this. I was hoping we had a more clean way to do it that doesn't teleport them back every second or two.

File: code/datums/components/leash.dm (hunk 3 above)
Comment (Mothblocks) [reply code/datums/components/leash.dm:170]: Yeah, eventually something is going to regress and it's not really any code to leave around

File: code/datums/components/leash.dm [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+/obj/effect/spawner/debug_leash/Initialize()
```
Comment (Mothblocks) [inline code/datums/components/leash.dm:172]:
```suggestion
/obj/effect/spawner/debug_leash/Initialize(mapload)
```

**Pull request merged**